File: Python/ByTier/A/0720_A_Closing_ceremony.py
# Problem: CF 720 A - Closing ceremony
# https://codeforces.com/contest/720/problem/A

"""
Algorithm: Greedy with Binary Search
Approach:
- We simulate the seating process by iterating through seats in a specific order.
- For each seat (i, j), we check if there's a person from either queue who can reach it.
- The queues are:
  - Queue 1 at (0, 0): people who can walk from (0, 0) to (i, j) with distance i + j
  - Queue 2 at (0, m+1): people who can walk from (0, m+1) to (i, j) with distance i + m + 1 - j
- For each seat, we use binary search to find the smallest stamina value that can cover the distance.
- If a person is found, they are removed from the queue.
- If no one is found for any seat, return "NO".

Time Complexity: O(n * m * log(n * m))
Space Complexity: O(n * m)
"""

import bisect
import sys

input = lambda: sys.stdin.readline().rstrip()

I = lambda: int(input())
MI = lambda: map(int, input().split())
GMI = lambda: map(lambda x: int(x) - 1, input().split())
LI = lambda: list(MI())
TI = lambda: tuple(MI())
LGMI = lambda: list(GMI())
YN = lambda x: print("YES" if x else "NO")
mod = 1000000007
mod2 = 998244353

tcn = 1
for _tcn_ in range(tcn):
    n, m = MI()
    n1, *nums1 = MI()  # nums1: list of staminas for people at (0,0)
    n2, *nums2 = MI()  # nums2: list of staminas for people at (0,m+1)
    nums1.sort()       # sort staminas for binary search
    nums2.sort()
    ans = True         # Flag to indicate if distribution is possible
    for i in range(n, 0, -1):  # Iterate from last row to first
        for j in range(1, m + 1):  # Iterate from first column to last
            # Calculate distances for both queues
            dist1 = i + j                # distance for queue1 (0,0)
            dist2 = i + m + 1 - j        # distance for queue2 (0,m+1)
            
            # Find the smallest stamina value >= dist1 in nums1
            iter1 = bisect.bisect_left(nums1, dist1)
            # Find the smallest stamina value >= dist2 in nums2
            iter2 = bisect.bisect_left(nums2, dist2)
            
            # If there is a person in queue1 who can cover distance1, assign to them
            if iter1 < len(nums1):
                nums1.pop(iter1)
            # Else if there is a person in queue2 who can cover distance2, assign to them
            elif iter2 < len(nums2):
                nums2.pop(iter2)
            # If no one can cover either distance, it's impossible
            else:
                ans = False
                break
        if ans is False:
            break
    YN(ans)


# https://github.com/VaHiX/CodeForces/

File: CPP/ByRound/0720/0720_A_Closing_ceremony.cpp
// Problem: CF 720 A - Closing ceremony
// https://codeforces.com/contest/720/problem/A

/*
 * Problem: Closing Ceremony
 * 
 * Algorithm/Technique: Greedy + Sorting + Two Pointers
 * 
 * Time Complexity: O(n * m * log(n * m))
 * Space Complexity: O(n * m)
 * 
 * Approach:
 * - We model the seating arrangement as a grid of n rows and m columns.
 * - People are divided into two queues:
 *   - Queue 1 at (0, 0) with k people having stamina values.
 *   - Queue 2 at (0, m+1) with (n*m - k) people having stamina values.
 * - We want to assign each seat to a person such that the person's stamina 
 *   is sufficient for the Manhattan distance to reach the seat.
 * - Greedy assignment is used:
 *   - Process seats row by row from bottom to top.
 *   - For each seat (i,j), compute the minimum distance from both queues.
 *   - Assign the seat to a person with sufficient stamina.
 *   - If no one is available, return "NO".
 * 
 * Key observations:
 * - For seat (i, j), queue 1 (0,0) distance is (i + j).
 * - For seat (i, j), queue 2 (0, m+1) distance is (i + (m+1-j)).
 * - Sorting stamina arrays lets us use efficient binary search.
 * - Use lower_bound to find the smallest stamina value that satisfies the required distance.
 */

#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

int n, m, k, x;
vector<int> a, b;

int main() {
  cin >> n >> m >> k;
  // Read stamina values for people at (0,0)
  for (int i = 0; i < k; i++)
    cin >> x, a.push_back(x);
  // Read number of people at (0, m+1) and then their stamina values
  cin >> x;
  for (int i = 0; i < n * m - k; i++)
    cin >> x, b.push_back(x);
  
  // Sort stamina values for efficient matching
  sort(a.begin(), a.end()), sort(b.begin(), b.end());
  
  // Process seats in a specific order (bottom to top, left to right)
  for (int i = n; i >= 1; i--)
    for (int j = 1; j <= m; j++) {
      // Find minimum stamina needed for seat (i,j)
      // From queue 1 (0,0): distance = i + j
      // From queue 2 (0,m+1): distance = i + m + 1 - j
      auto p = lower_bound(a.begin(), a.end(), i + j);  // Check queue 1
      auto q = lower_bound(b.begin(), b.end(), i + m + 1 - j);  // Check queue 2
      
      // Try to assign to queue 1
      if (p != a.end()) {
        a.erase(p);  // Remove used stamina value
      } else if (q != b.end()) {
        b.erase(q);  // Remove used stamina value
      } else {
        // No one can reach this seat
        return cout << "NO", 0;
      }
    }
    
  cout << "YES";
}


// https://github.com/VaHiX/CodeForces/